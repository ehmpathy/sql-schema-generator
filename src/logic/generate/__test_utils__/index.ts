export * from './databaseConnection';
export * from './createTablesForEntity';
export * from './dropTablesForEntity';
