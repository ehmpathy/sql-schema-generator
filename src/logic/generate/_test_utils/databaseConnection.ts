import mysql, { Connection, ConnectionOptions } from 'mysql2/promise';

export const getDatabaseConnection = async (): Promise<Connection> => {
  const dbConfig: ConnectionOptions = {
    // hardocded since this is just the integration test db config; settings configured in the provisioning/.../docker.file
    host: 'localhost',
    user: 'root',
    password: 'a-secure-password', // its just a docker integration test db, no worries about hardcoding this
    database: 'superimportantdb',
    port: 12821,
  };
  const connection = await mysql.createConnection(dbConfig);
  return connection;
};

export { Connection as DatabaseConnection };
