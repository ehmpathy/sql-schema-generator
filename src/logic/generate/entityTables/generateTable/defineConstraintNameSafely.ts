import { toHashShake256Sync } from '../../../../__nonpublished_modules__/hash-fns/toHashShake256Sync';

const POSTGRES_RESOURCE_NAME_LEN_LIMIT = 63;

/**
 * .what = determines an intuitive yet safe constraint name
 * .why = postgres has a 63 char resource name limit
 * .how = include a hash if the ideal name would be too long
 */
export const defineConstraintNameSafely = (input: {
  tableName: string;
  constraintName: string;
}): string => {
  const idealConstraintName = [input.tableName, input.constraintName].join('_');
  if (idealConstraintName.length <= POSTGRES_RESOURCE_NAME_LEN_LIMIT)
    return idealConstraintName;
  const hash = toHashShake256Sync(input.tableName, 4);
  const tableNameLimit =
    POSTGRES_RESOURCE_NAME_LEN_LIMIT -
    hash.length -
    input.constraintName.length -
    2;
  const safeConstraintName = [
    input.tableName.slice(0, tableNameLimit),
    hash,
    input.constraintName,
  ].join('_');
  return safeConstraintName;
};
