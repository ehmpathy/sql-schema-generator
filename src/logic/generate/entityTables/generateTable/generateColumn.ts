import { DataTypeName, Property } from '../../../../types';
import { extractMysqlTypeDefinitionFromProperty } from '../../utils/extractMysqlTypeDefinitionFromProperty';

/*
  // TODO: generalize w/ adapter pattern to other languages
*/
const textTypes = [
  DataTypeName.CHAR,
  DataTypeName.VARCHAR,
  DataTypeName.TINYTEXT,
  DataTypeName.TEXT,
  DataTypeName.MEDIUMTEXT,
  DataTypeName.LONGTEXT,
  DataTypeName.ENUM,
];
export const generateColumn = ({ columnName, property }: { columnName: string, property: Property }) => {

  // define default over-rides
  let defaultNullRelation = (property.nullable) ? 'DEFAULT NULL' : ''; // if nullable, then the "default" default is "null"; else, nothing
  if (property.type.name === DataTypeName.TEXT) defaultNullRelation = ''; // if its varchar, then even if nullable don't show default as null; // TODO: discover why SHOW CREATE has this exception

  // define the full column definition
  const modifiers = [
    `\`${columnName}\``,

    extractMysqlTypeDefinitionFromProperty({ property }), // e.g., property => bigint(20)

    (textTypes.includes(property.type.name)) ? 'COLLATE utf8_bin' : '', // coalate if its a string type; collate to Case Sensitive by default

    (property.nullable) ? '' : 'NOT NULL', // if not nullable, then NOT NULL

    (property.default) ? `DEFAULT ${property.default}` : defaultNullRelation, // if default is set, use it; otherwise, no default

    (property.comment) ? `COMMENT '${property.comment}'` : '', // empty string if comment not defined

    (columnName === 'id') ? 'AUTO_INCREMENT' : '',
  ];
  return modifiers
    .filter(modifier => !!modifier) // filter out null
    .join(' '); // merge into column
};
