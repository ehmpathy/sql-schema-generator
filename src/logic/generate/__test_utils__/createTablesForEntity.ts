import { Entity } from '../../../types';
import { generateEntityTables } from '../entityTables/generateEntityTables';
import { DatabaseConnection } from './databaseConnection';

export const createTablesForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  const tables = await generateEntityTables({ entity });

  // create the static table
  await dbConnection.query({ sql: tables.static.sql });

  // drop the version table, if exists
  if (tables.version) {
    await dbConnection.query({ sql: tables.version.sql });
  }

  // drop the cvp table, if exists
  if (tables.currentVersionPointer) {
    await dbConnection.query({ sql: tables.currentVersionPointer.sql });
  }

  // create each mapping table
  await Promise.all(tables.mappings.map((mappingTable) => dbConnection.query({ sql: mappingTable.sql })));
};
