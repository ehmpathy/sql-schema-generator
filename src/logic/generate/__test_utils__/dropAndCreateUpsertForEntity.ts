import { Entity } from '../../../types';
import { generateEntityUpsert } from '../entityFunctions/generateEntityUpsert';
import { DatabaseConnection } from './databaseConnection';
import { provisionGetFromDelimiterSplitStringFunction } from './provisionGetFromDelimiterSplitStringFunction';

export const dropAndCreateUpsertFunctionForEntity = async ({
  entity,
  dbConnection,
}: {
  entity: Entity;
  dbConnection: DatabaseConnection;
}) => {
  // if has array properties, provision the array access function
  if (Object.values(entity.properties).some((prop) => !!prop.array)) {
    await provisionGetFromDelimiterSplitStringFunction({ dbConnection });
  }

  // drop and create the upsert
  const { name, sql: upsertSql } = generateEntityUpsert({ entity });
  await dbConnection.query({ sql: `DROP FUNCTION IF EXISTS ${name}` });
  await dbConnection.query({ sql: upsertSql });
};
