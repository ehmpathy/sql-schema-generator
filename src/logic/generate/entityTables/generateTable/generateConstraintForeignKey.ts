import { Property } from '../../../../domain';
import { defineConstraintNameSafely } from './defineConstraintNameSafely';

export const generateConstraintForeignKey = ({
  index,
  tableName,
  columnName,
  property,
}: {
  index: number;
  tableName: string;
  columnName: string;
  property: Property;
}) => {
  const constraintName = `${tableName}`;
  const constraintSql = `
CONSTRAINT ${defineConstraintNameSafely({
    tableName,
    constraintName: `fk${index}`,
  })} FOREIGN KEY (${columnName}) REFERENCES ${property.references!} (id)
  `.trim();

  const indexSql = `
CREATE INDEX ${defineConstraintNameSafely({
    tableName,
    constraintName: `fk${index}_ix`,
  })} ON ${tableName} USING btree (${columnName});
  `.trim();

  return {
    index: indexSql,
    constraint: constraintSql,
  };
};
