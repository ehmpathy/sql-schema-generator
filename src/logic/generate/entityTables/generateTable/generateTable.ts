import { Property } from '../../../../domain';
import { generateColumn } from './generateColumn';
import { generateConstraintForeignKey } from './generateConstraintForeignKey';

export const generateTable = ({
  tableName,
  properties,
  unique,
}: {
  tableName: string;
  properties: { [index: string]: Property };
  unique: string[];
}) => {
  // 0. validate input
  if (unique.length === 0) {
    throw new Error('must have atleast one unique property; otherwise, idempotency cant be enforced');
  }

  // define sql per column
  const columnSqls = Object.entries(properties).map((entry) =>
    generateColumn({ columnName: entry[0], property: entry[1] }),
  );

  // define primary key
  const primaryKeySql = `CONSTRAINT ${tableName}_pk PRIMARY KEY (id)`;

  // define unique index
  const uniqueConstraintSql = `CONSTRAINT ${tableName}_ux1 UNIQUE (${unique.join(', ')})`; // unique key definition; required since it is required for idempotency

  // define foreign keys
  const foreignKeySqls = Object.entries(properties)
    .filter((entry) => !!entry[1].references)
    .map((entry, index) =>
      generateConstraintForeignKey({ index, columnName: entry[0], tableName, property: entry[1] }),
    );
  const foreignKeyIndexSqls = foreignKeySqls.map((sqls) => sqls.index);
  const foreignKeyConstraintSqls = foreignKeySqls.map((sqls) => sqls.constraint);

  // define check constraints
  const checkConstraintSqls = Object.entries(properties)
    .filter((entry) => !!entry[1].check)
    .map((entry) => {
      return `CONSTRAINT ${tableName}_${entry[0]}_check CHECK ${entry[1].check!.replace(/\$COLUMN_NAME/g, entry[0])}`;
    });

  // 2. define the lines of content of the table
  const contents = [
    ...columnSqls, // all of the columns (one per property)
    primaryKeySql,
    uniqueConstraintSql,
    ...foreignKeyConstraintSqls, // constraints defined for FKs
    ...checkConstraintSqls, // constraints defined for checks
  ];

  // define the indexes we're creating
  const indexes = [
    // NOTE: pk index is auto created
    // NOTE: ux index is auto created
    ...foreignKeyIndexSqls, // indexes defined for FKs
  ];

  // 2. generate the create table sql
  const sql = `
CREATE TABLE ${tableName} (
  ${contents.join(',\n  ')}
);
${indexes.join('\n')}
  `.trim();

  // 3. return the sql
  return sql;
};
