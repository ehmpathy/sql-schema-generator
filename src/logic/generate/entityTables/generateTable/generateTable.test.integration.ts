import { DatabaseConnection, getDatabaseConnection } from '../../../../__test_utils__/databaseConnection';
import { getShowCreateTable } from '../../../../__test_utils__/getShowCreateTable';
import { DataType, DataTypeName, Property } from '../../../../types';
import { prop } from '../../../define';
import { generateTable } from './generateTable';

/*
  we test an example of every variation against the database to ensure we are defining valid sql
*/
describe('generateTable', () => {
  let dbConnection: DatabaseConnection;
  beforeAll(async () => {
    dbConnection = await getDatabaseConnection();
  });
  afterAll(async () => {
    await dbConnection.end();
  });
  const testTableIsCreateable = async ({ createTableSql }: { createTableSql: string }) => {
    await dbConnection.query({ sql: 'DROP TABLE IF EXISTS generate_table_test CASCADE;' });
    await dbConnection.query({ sql: 'DROP TABLE IF EXISTS generate_table_test_referenced CASCADE;' });
    await dbConnection.query({
      sql: `
      CREATE TABLE generate_table_test_referenced (
        id BIGINT PRIMARY KEY
      )
    `,
    });
    await dbConnection.query({ sql: createTableSql });
  };
  const getShowCreateNow = async () => getShowCreateTable({ table: 'generate_table_test', dbConnection });
  it('can create a table with basic column definition, w/ same syntax as from SHOW CREATE TABLE', async () => {
    const pk = new Property({
      type: new DataType({
        name: DataTypeName.BIGINT,
      }),
    });
    const reference = new Property({
      type: new DataType({
        name: DataTypeName.BIGINT,
      }),
      references: 'generate_table_test_referenced',
    });
    const status = prop.ENUM(['QUEUED', 'ATTEMPTED', 'FULFILLED']);
    const createTableSql = await generateTable({
      tableName: 'generate_table_test',
      properties: { id: pk, reference_id: reference, second_reference_id: reference, status },
      unique: ['reference_id'],
    });
    await testTableIsCreateable({ createTableSql });

    // check syntax is the same as that returned by SHOW CREATE TABLE
    const createTableSqlFound = await getShowCreateNow();
    expect(createTableSqlFound).toEqual(createTableSql); // should be the exact string

    // snapshot to save an example
    expect(createTableSql).toMatchSnapshot();
  });
});
