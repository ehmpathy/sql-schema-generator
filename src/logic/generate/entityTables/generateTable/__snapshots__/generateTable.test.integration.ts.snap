// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateTable can create a table with basic column definition, w/ same syntax as from SHOW CREATE TABLE 1`] = `
"CREATE TABLE generate_table_test (
  id bigint NOT NULL,
  reference_id bigint NOT NULL,
  second_reference_id bigint NOT NULL,
  status varchar NOT NULL,
  CONSTRAINT generate_table_test_pk PRIMARY KEY (id),
  CONSTRAINT generate_table_test_ux1 UNIQUE (reference_id),
  CONSTRAINT generate_table_test_fk0 FOREIGN KEY (reference_id) REFERENCES generate_table_test_referenced (id),
  CONSTRAINT generate_table_test_fk1 FOREIGN KEY (second_reference_id) REFERENCES generate_table_test_referenced (id),
  CONSTRAINT generate_table_test_status_check CHECK (status IN ('QUEUED', 'ATTEMPTED', 'FULFILLED'))
);
CREATE INDEX generate_table_test_fk0_ix ON generate_table_test USING btree (reference_id);
CREATE INDEX generate_table_test_fk1_ix ON generate_table_test USING btree (second_reference_id);"
`;
